File: dynamic_programming/subarray_and_subsequence/1143_longest_common_subsequence.cc
#include <iostream>
#include <string>
#include <vector>

// 2024/01/13: 不看答案自己做出来. 难点在与状态转移方程的推导.

int LongestCommonSubsequence(const std::string& text1,
                             const std::string& text2) {
  const int m = text1.size();
  const int n = text2.size();
  std::vector<std::vector<int>> dp(m + 1, std::vector<int>(n + 1, 0));
  for (int i = 1; i <= m; ++i) {
    for (int j = 1; j <= n; ++j) {
      if (text1[i - 1] == text2[j - 1]) {
        dp[i][j] = dp[i - 1][j - 1] + 1;
      } else {
        dp[i][j] = std::max(dp[i - 1][j], dp[i][j - 1]);
      }
      std::cout << "dp[" << i << "][" << j << "] = " << dp[i][j] << std::endl;
    }
  }
  return dp[m][n];
}

int main(int argc, char* argv[]) {
  std::string text1 = "abcde";
  std::string text2 = "ace";
  std::cout << LongestCommonSubsequence(text1, text2) << std::endl;
  return 0;
}

File: dynamic_programming/string_edit/1143_longest_common_subsequence.py
# 2026/06/13: 不看答案自己实现, 用时 19min.
def longest_common_subsequence(text1: str, text2: str) -> int:
    # dp[i, j] 表示 text1 以 i 结尾, text2 以 j 结尾的最长公共字串的长度.
    m, n = len(text1), len(text2)
    dp = [[0] * n for _ in range(m)]
    for i in range(m):
        for j in range(n):
            if i == 0 and j == 0:
                dp[i][j] = int(text1[i] == text2[j])
            elif i == 0:
                dp[i][j] = max(int(text1[i] == text2[j]), dp[i][j - 1])
            elif j == 0:
                dp[i][j] = max(int(text1[i] == text2[j]), dp[i - 1][j])
            else:
                dp[i][j] = max(
                    int(text1[i] == text2[j]) + dp[i - 1][j - 1],
                    dp[i - 1][j],
                    dp[i][j - 1],
                )
    return dp[-1][-1]


text1 = "abcde"
text2 = "ace"
print(longest_common_subsequence(text1, text2))
